File: ts/src/allowlist.ts
interface NormalizedPattern {
  type: 'exact' | 'wildcard_prefix' | 'bare_domain';
  value: string;
}

export class Allowlist {
  private patterns: NormalizedPattern[];

  constructor(patterns: string[]) {
    this.patterns = [];
    for (const p of patterns) {
      const trimmed = p.trim();
      if (!trimmed) {
        continue;
      }
      // Reject patterns with dangerous characters
      if (/[\[\]?]/.test(trimmed)) {
        console.warn(`Rejecting dangerous allowlist pattern: ${trimmed}`);
        continue;
      }
      const lower = trimmed.toLowerCase();
      if (lower.startsWith('*@')) {
        // Wildcard prefix: *@domain.tld
        this.patterns.push({
          type: 'wildcard_prefix',
          value: lower.slice(2),
        });
      } else if (lower.includes('@')) {
        // Exact match: user@domain.tld
        this.patterns.push({
          type: 'exact',
          value: lower,
        });
      } else {
        // Bare domain: domain.tld
        this.patterns.push({
          type: 'bare_domain',
          value: lower,
        });
      }
    }
  }

  isAllowed(email: string): boolean {
    const emailLower = email.toLowerCase();
    return this.patterns.some((pattern) => this.matchesPattern(pattern, emailLower));
  }

  private matchesPattern(pattern: NormalizedPattern, email: string): boolean {
    switch (pattern.type) {
      case 'exact':
        return email === pattern.value;
      case 'wildcard_prefix':
        return email.endsWith(`@${pattern.value}`);
      case 'bare_domain':
        return email.endsWith(`@${pattern.value}`);
    }
  }
}

File: rust/src/allowlist.rs
#[derive(Debug, Clone)]
enum PatternType {
    Exact(String),
    WildcardPrefix(String),
    BareDomain(String),
}

pub struct Allowlist {
    patterns: Vec<PatternType>,
}

impl Allowlist {
    pub fn new(patterns: &[String]) -> Self {
        let normalized: Vec<PatternType> = patterns
            .iter()
            .map(|p| p.trim().to_string())
            .filter(|p| !p.is_empty())
            .filter_map(|p| {
                let lower = p.to_lowercase();
                // Reject patterns with dangerous characters
                if lower.contains('[') || lower.contains(']') || lower.contains('?') {
                    eprintln!("Rejecting dangerous allowlist pattern: {}", p);
                    return None;
                }
                Some((p, lower))
            })
            .map(|(_, lower)| {
                if let Some(domain) = lower.strip_prefix("*@") {
                    // Wildcard prefix: *@domain.tld
                    PatternType::WildcardPrefix(domain.to_string())
                } else if lower.contains('@') {
                    // Exact match: user@domain.tld
                    PatternType::Exact(lower)
                } else {
                    // Bare domain: domain.tld
                    PatternType::BareDomain(lower)
                }
            })
            .collect();
        Self { patterns: normalized }
    }

    pub fn is_allowed(&self, email: &str) -> bool {
        let email_lower = email.to_lowercase();
        self.patterns
            .iter()
            .any(|pattern| self.matches_pattern(pattern, &email_lower))
    }

    fn matches_pattern(&self, pattern: &PatternType, email: &str) -> bool {
        match pattern {
            PatternType::Exact(p) => email == p,
            PatternType::WildcardPrefix(domain) => email.ends_with(&format!("@{}", domain)),
            PatternType::BareDomain(domain) => email.ends_with(&format!("@{}", domain)),
        }
    }
}
